File: DEMO/novak.py
import hashlib
import json
import time

# -----------------------------------------
# SHA-256 helper (normalized representation)
# -----------------------------------------
def _h(value):
    """Hash any Python object deterministically."""
    if isinstance(value, (dict, list)):
        value = json.dumps(value, sort_keys=True).encode()
    elif isinstance(value, str):
        value = value.encode()
    else:
        value = str(value).encode()
    return hashlib.sha256(value).hexdigest()

# -----------------------------------------
# NOVAK GENESIS HASH
# -----------------------------------------
GENESIS_HASH = _h("NOVAK-GENESIS")

# -----------------------------------------
# NIPS: Input Verification
# -----------------------------------------
def verify_nips(inputs, schema):
    # Check required fields exist
    for field in schema:
        if field not in inputs:
            return False
    
    # Check attestation flag
    if not inputs.get("attested", False):
        return False

    return True

# -----------------------------------------
# HARMONEE: Execution Receipt
# -----------------------------------------
def harmonee_receipt(rule, inputs, output):
    receipt_obj = {
        "rule_id": rule["id"],
        "input_hash": _h(inputs),
        "output_hash": _h(output),
        "actor": "NOVAK_KERNEL",
        "timestamp": int(time.time())
    }
    return _h(receipt_obj), receipt_obj

# -----------------------------------------
# REVELATION: Global recursive hash chain
# -----------------------------------------
def revelation_chain(previous_hash, receipt_obj):
    chain_obj = {
        "prev_hash": previous_hash,
        "receipt": receipt_obj
    }
    return _h(chain_obj)

# -----------------------------------------
# Equal Execution Law (EEL)
# -----------------------------------------
def equal_execution_check(rule, inputs, output, reference_table):
    key = _h({"rule": rule["id"], "inputs": inputs})

    if key in reference_table:
        expected = reference_table[key]
        if expected != output:
            return {
                "violation": True,
                "expected": expected,
                "actual": output,
                "justification_required": True
            }
    else:
        reference_table[key] = output

    return {"violation": False}

# -----------------------------------------
# NOVAK EXECUTION KERNEL
# -----------------------------------------
def novak_execute(rule, inputs, schema, reference_table, prev_chain_hash=GENESIS_HASH):

    # 1. NIPS — Verified inputs only
    if not verify_nips(inputs, schema):
        raise ValueError("NIPS Violation: Inputs invalid or not attested.")

    # 2. Deterministic rule logic
    cond = rule["logic"]["if"]
    if inputs[cond["field"]] >= cond["value"]:
        output = rule["logic"]["then"]
    else:
        output = rule["logic"]["else"]

    # 3. HARMONEE receipt
    receipt_hash, receipt_obj = harmonee_receipt(rule, inputs, output)

    # 4. REVELATION global chain
    chain_hash = revelation_chain(prev_chain_hash, receipt_obj)

    # 5. Equal Execution Law
    eq_check = equal_execution_check(rule, inputs, output, reference_table)

    return {
        "output": output,
        "receipt_hash": receipt_hash,
        "receipt_obj": receipt_obj,
        "chain_hash": chain_hash,
        "equal_execution": eq_check
    }

# -----------------------------------------
# SIMPLE TEST EXECUTION (DEMO)
# -----------------------------------------

if __name__ == "__main__":

    # Example rule
    rule = {
        "id": "RULE-V1",
        "logic": {
            "if": {"field": "score", "value": 70},
            "then": {"approved": True, "level": "A"},
            "else": {"approved": False, "level": "F"}
        }
    }

    # Example schema required fields
    schema = ["score", "attested"]

    # Example inputs (valid)
    inputs = {
        "score": 85,
        "attested": True
    }

    # Equal execution reference table
    reference_table = {}

    # Execute NOVAK
    result = novak_execute(
        rule=rule,
        inputs=inputs,
        schema=schema,
        reference_table=reference_table
    )

    # Pretty print the result
    print("\n=== NOVAK EXECUTION RESULT ===")
    print(json.dumps(result, indent=4))

File: reference/javascript/Novak.js
// NOVAK Protocol Demo
// Matthew Novak – 2025
// Implements: NIPS, HARMONEE, REVELATION, Equal Execution Law

// SHA-256 function (browser + Node compatible)
async function sha256(text) {
  const enc = new TextEncoder().encode(text);
  const hashBuffer = await crypto.subtle.digest("SHA-256", enc);
  return [...new Uint8Array(hashBuffer)]
    .map(b => b.toString(16).padStart(2, "0"))
    .join("");
}

// ----------------------
// 1) NIPS — Input Proof
// ----------------------
async function NIPS(input) {
  if (!input) throw new Error("NIPS FAIL: Missing input");
  if (typeof input !== "object") throw new Error("NIPS FAIL: Must be object");

  // Simulated signature + schema check
  if (!input.attested) throw new Error("NIPS FAIL: Input not attested");
  if (!input.schemaVersion) throw new Error("NIPS FAIL: Bad schema");

  return {
    input,
    inputHash: await sha256(JSON.stringify(input))
  };
}

// ----------------------
// 2) HARMONEE — Execution Receipt
// ----------------------
async function HARMONEE(ruleId, inputHash, output, actor="system") {
  const outputHash = await sha256(JSON.stringify(output));
  const timestamp = Date.now().toString();

  const receiptString = `${ruleId}|${inputHash}|${outputHash}|${actor}|${timestamp}`;
  const receiptHash = await sha256(receiptString);

  return {
    ruleId,
    inputHash,
    outputHash,
    actor,
    timestamp,
    receiptHash
  };
}

// ----------------------
// 3) REVELATION — Global Chain
// ----------------------
async function REVELATION(prev, receiptHash) {
  const chainHash = await sha256(prev + "|" + receiptHash);
  return chainHash;
}

// ----------------------
// 4) NOVAK Equal Execution Law
// ----------------------
function equalExecutionCheck(D1, D2, O1, O2) {
  const inputsEqual = JSON.stringify(D1) === JSON.stringify(D2);
  const outputsEqual = JSON.stringify(O1) === JSON.stringify(O2);

  if (inputsEqual && !outputsEqual) {
    throw new Error("NOVAK VIOLATION: Identical inputs produced different outputs");
  }

  return true;
}

// ----------------------
// DEMO RUN
// ----------------------
async function demoNOVAK() {
  console.log("=== NOVAK DEMO START ===");

  // 1. NIPS input
  const input = {
    claimant: "12345",
    condition: "hearing_loss",
    attested: true,
    schemaVersion: 1
  };

  const nips = await NIPS(input);

  // 2. Fake rule engine
  function ruleEngine(input) {
    return {
      approved: true,
      rating: 10
    };
  }

  const output = ruleEngine(input);

  // 3. Equal execution check
  equalExecutionCheck(input, input, output, output);

  // 4. HARMONEE receipt
  const receipt = await HARMONEE("RULE-VA-2025", nips.inputHash, output);

  // 5. REVELATION chain
  const chain = await REVELATION("GENESIS", receipt.receiptHash);

  console.log({ nips, receipt, chain });
  console.log("=== NOVAK DEMO COMPLETE ===");
}

demoNOVAK();
